### PR TITLE
ISO-TP: new ISO-TP implementation

### PR DESCRIPTION
**Pull request by garethpotter**
- single public context for send and receive; no more unbinding to send a message
- only supports ISO-TP fixed addressing at present
- broadly async approach, but sync receive supported for familiarity/compatibility
- conformance tests ported from existing implementation

### CODE REVIEW

**Comment by martinjaeger** [comment]
Thanks @garethpotter. Could you please rebase on main to resolve the conflicts?

File: src/isotp_fast.c [hunk 1]
```diff
@@ -0,0 +1,1058 @@
...
+}
+
+int isotp_fast_bind(struct isotp_fast_ctx *ctx, const struct device *can_dev,
+                    const isotp_fast_msg_id my_addr, const struct isotp_fast_opts *opts,
+                    isotp_fast_recv_callback_t recv_callback, void *recv_cb_arg,
+                    isotp_fast_recv_error_callback_t recv_error_callback,
+                    isotp_fast_send_callback_t sent_callback)
+{
+    sys_slist_init(&isotp_send_ctx_list);
+    sys_slist_init(&isotp_recv_ctx_list);
```
Comment (martinjaeger) [inline src/isotp_fast.c:850]: Is it safe to re-initialize the `slist` with every bind? And does this still allow to have multiple binds?

Comment (garethpotter) [reply src/isotp_fast.c:850]: I think this did cross my mind at one point actually, but I never got round to doing anything about it. The lists should probably sit on the `isotp_fast_context` itself.

**Comment by garethpotter** [comment]
> I have added a few commits to re-structure the tests and fix a bug. Now it is passing CI.

Noted. Thanks.

> We should use the terms CAN ID vs. ISO-TP address consistently. `isotp_fast_msg_id` is actually a CAN ID and variables of that type should not be named `my_addr` or `recipient_addr`. ISO-TP uses the terms source address and target address for the 1-byte section in the CAN ID.

Yes, fair enough. I was trying to make it make more sense to me as I worked on it but it would definitely be better to change it back for consistency.

> Do we need thread synchronization to avoid race conditions when accessing the global `slist`s?

Possibly, but if these move to the `isotp_fast_context` then that becomes slightly less problematic.

> Should we replace `isotp_fast_context` with `isotp_fast_binding` to avoid confusion with `sctx` and `rctx`?

I prefer `context`, because this is the only user-facing structure (replacing the existing send/receive contexts), but if you think there is real potential for confusion then let's rename it.

**Comment by martinjaeger** [comment]
> I prefer `context`, because this is the only user-facing structure (replacing the existing send/receive contexts), but if you think there is real potential for confusion then let's rename it.

Ok, fine with me.

**Comment by garethpotter** [comment]
I'm not quite sure what to do regarding locking yet. I agree that there ought to be something in place, but I'm not entirely sure where. As it is, we have a global buffer pool and global memory slabs for the contexts and access to these is not protected by locks. I did contemplate moving the pools and slabs to the `isotp_fast_context` (which would at least reduce the need for locking, if not eliminate it), but while defining memory slabs without the macro is easy, it is less easy to do that for pools, so I have shied away that for now.

File: src/can.c [hunk 2]
```diff
@@ -627,7 +778,7 @@ static void thingset_can_thread()
     }
 
     while (true) {
-        thingset_can_process_inst(&ts_can_single, K_FOREVER);
+        k_sleep(K_FOREVER);
```
Comment (martinjaeger) [inline src/can.c:781]: Should the `process` function not be kept if `CONFIG_ISOTP_FAST` is not used?

Also when it's used we don't need the overhead and additional stack of the thread anymore, if it sits there and sleeps forever. We should use `SYS_INIT` instead.

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Thread sync should be fine now with the `slist`s moved to the contexts, as we will only have one ISO-TP instance when used with ThingSet anyway.

Only got one remaining comment.

**Review by martinjaeger — APPROVED** [review]
Thanks for the update.

Let's go ahead and merge this exciting new feature.

**Pull request merged**